### PR TITLE
Inline code intel extensions in Firefox addon

### LINKED ISSUE #11369: Inline code intel extensions in the Firefox extension
For context, see [RFC 173](https://docs.google.com/document/u/1/d/1YN9xdRD7RhAkCrR2RYbIQpYrC_wrVNteuzp_nCguuoI/edit). Outline of steps needed to achieve this:

- [ ] Include built [code-intel-extensions](https://github.com/sourcegraph/code-intel-extensions) in Firefox extension assets
    - We need the extension's "manifest" (package.json), and the built JS bundle
    - We may actually get away with only packaging three extensions: the Go and TypeScript extensions, and a generic extension for all other languages. This is the [template extension](https://sourcegraph.com/github.com/sourcegraph/code-intel-extensions/-/tree/extensions/template), from which individual (Cpp, Lua, Python, ...) extensions are generated at build time and individually published to the extensions registry.
- [ ] Make sure we only try to activate extensions that are inlined in browser extension assets.
    - One way to do this would be to specify an optional [extensions service](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@1a4ccf3649f0af18337f96ae4487fe28505fa93a/-/blob/shared/src/api/client/services/extensionsService.ts#L58-177) implementation on [`PlatformContext`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@1a4ccf3649f0af18337f96ae4487fe28505fa93a/-/blob/shared/src/platform/context.ts#L50-56), which would be used when [instantiating services](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@1a4ccf3649f0af18337f96ae4487fe28505fa93a/-/blob/shared/src/api/client/services.ts#L48) if present. Querying active extensions from settings, sideloading extensions, fetching extensions from the registry will not make sense in the Firefox extension.
- [ ] Load extensions from inlined assets
    - I think if there are only a handful of extensions, we could simply `require`/`import` the manifests, and thus include them in our main JS bundles by default. This could be done in the alternative `ExtensionsService` implementation mentioned above.
    - To load the extension bundles, I think we should pass URLs to the bundles created using [runtime.getURL()](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/getURL), which, given a relative path to a file packaged within the extension, returns a fully qualified path (eg. `extensions/go/bundle.js` -> `moz-extension://SOME_GUID/extensions/go/bundle.js`). These paths will have to be created outside of the worker scope (as browser APIs are not available in the worker). The correct place for this would be the implementation of [`PlatformContext.getScriptURLForExtension()`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@1a4ccf3649f0af18337f96ae4487fe28505fa93a/-/blob/browser/src/shared/platform/context.ts#L144-155). These fully-qualified URLs can then be seamlessly used in [`importScripts()`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@1a4ccf3649f0af18337f96ae4487fe28505fa93a/-/blob/shared/src/api/extension/api/extensions.ts#L41:1) calls in the extension host worker.

### LINKED ISSUE #11371: Remove script-src blob: from the CSP of our Firefox extension
Prerequisite: #11369.

`script-src blob:` breaches the following Mozilla [add-on policy](https://extensionworkshop.com/documentation/publish/add-on-policies/#development-practices):

> Add-ons must be self-contained and not load remote code for execution

CSP directives for the browser extension are set in [`manifest.spec.json`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@1a4ccf3/-/blob/browser/src/browser-extension/manifest.spec.json#L31:31). We could stop specifying them there, and add them at build time in [`writeManifest()`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@f3c0274598f5cba73256a8a55db75519be58a7c5/-/blob/browser/scripts/tasks.ts#L98-121) instead.

### PR DESCRIPTION
**Pull request by marekweb**
For now, this uses `sourcegraph/template` to apply code-intel to all supported languages.

Changes to the build script include:

- Remove `blob` from the Content Security Policy on Firefox (make it use the default CSP)
- Clone and build `code-intel-extensions` and copy it into the add-on build

Changes at run-time:

- When running in the Firefox addon, use a hardcoded list of extensions to load (current only `sourcegraph/template`)
- Load the extension locally (by fetching the manifest file `

Not included in this PR:

- generate individual extensions from code-intel-extensions for each supported languagge
- use settings to determine which extensions to enable/disable

Fix #11369 
Fix #11371

### CODE REVIEW

File: shared/src/api/extension/worker.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import ExtensionHostWorker from 'worker-loader?inline&name=extensionHostWorker.bundle.js!./main.worker.ts'
+import ExtensionHostWorker from 'worker-loader?name=extensionHostWorker.bundle.js&publicPath=js/!./main.worker.ts'
```
Comment (felixfbecker) [inline shared/src/api/extension/worker.ts:1]: This is a file in shared/, so it also runs in the webapp, where I'm not sure if this `publicPath` is correct?

Comment (marekweb) [reply shared/src/api/extension/worker.ts:1]: True, this needs to be checked

Comment (lguychard) [reply shared/src/api/extension/worker.ts]: I checked this in:
- The Chrome extension: worked like a charm
- Native integrations: worked well, but that was "by accident": `publicPath` is not used in native integrations, since we build the URL to the worker bundle [directly in `extensionHostFrame.html`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@d9905d3b073a084b3344707ee20bf3b60a4114ce/-/blob/browser/src/native-integration/extensionHostFrame.html?subtree=true#L14:32)
- Webapp: this did not work, the worker failed to load: `GET https://sourcegraph.test:3443/bitbucket.sgdev.org/SOURCEGRAPH/jsonrpc2/-/blob/js/extensionHostWorker.bundle.js 404`

I propose solving it this way: https://github.com/sourcegraph/sourcegraph/pull/11535/commits/987e90940c21891ec7a002e3f94be8b317e8e794 (I took the liberty of pushing to your branch, but feel free to tweak this if you feel it should be done differently)

As an FYI, to test this in native integrations, I:
- Ran the local dev environment with `./enterprise/dev/start.sh`
- Navigated to https://bitbucket.sgdev.org/plugins/servlet/sourcegraph (the Sourcegraph integration is already activated on bitbucket.sgdev.org, but it otherwise would've been [easy to set up](https://github.com/sourcegraph/bitbucket-server-plugin))
- Set `https://sourcegraph.test:3443` as Sourcegraph URL in the settings of the Bitbucket server plugin:
image
- Navigated to https://bitbucket.sgdev.org/projects/SOURCEGRAPH/repos/jsonrpc2/browse/async.go (that repo is also synced in the [default external services](https://github.com/sourcegraph/dev-private/blob/master/enterprise/dev/external-services-config.json#L49-L54) of the local dev environment)
- Verified that extensions were getting correctly activated, and that I got code intelligence.

File: shared/src/api/client/api/extensions.ts [hunk 2]
```diff
@@ -15,12 +15,15 @@ export class ClientExtensions {
...
-    constructor(private proxy: Remote<ExtExtensionsAPI>, extensionRegistry: ExtensionsService) {
+    constructor(private proxy: Remote<ExtExtensionsAPI>, extensionRegistry: IExtensionsService) {
         this.subscriptions.add(
             from(extensionRegistry.activeExtensions)
                 .pipe(startWith([] as ExecutableExtension[]), bufferCount(2, 1))
                 .subscribe(([oldExtensions, newExtensions]) => {
                     // Diff next state's activated extensions vs. current state's.
+                    if (!newExtensions) {
+                        newExtensions = oldExtensions
+                    }
```
Comment (lguychard) [inline shared/src/api/client/api/extensions.ts:26]: Why is this change necessary? `bufferCount()` should guarantee that we never emit a falsy `oldExtensions` (we might emit an empty array, but it will not be falsy).

File: browser/src/shared/platform/inlineExtensionsService.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                ] as ExecutableExtension[]
+        )
+
+    return from(requestPromise).pipe(shareReplay(1))
+}
+
+/**
+ * Manages the set of inline (bundled) extensions that are loaded. Only applicable to the Firefox browser addon.
+ *
+ * The inline extensions service loads extesions directly from the addon and does not load any code remotely.
```
Comment (lguychard) [inline browser/src/shared/platform/inlineExtensionsService.ts:50]:
```suggestion
 * The inline extensions service loads extensions directly from the add-on assets and does not load any code remotely.
```

File: browser/src/shared/platform/inlineExtensionsService.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                        scriptURL,
+                    },
+                ] as ExecutableExtension[]
+        )
+
+    return from(requestPromise).pipe(shareReplay(1))
+}
+
+/**
+ * Manages the set of inline (bundled) extensions that are loaded. Only applicable to the Firefox browser addon.
```
Comment (lguychard) [inline browser/src/shared/platform/inlineExtensionsService.ts:48]:
```suggestion
 * Manages the set of inline (bundled) extensions that are loaded. Only applicable to the Firefox browser add-on.
```

File: browser/src/shared/platform/inlineExtensionsService.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    return {
+        manifestURL: browser.extension.getURL(`extensions/${extensionName}/package.json`),
+        scriptURL: browser.extension.getURL(`extensions/${extensionName}/extension.js`),
+    }
+}
+
+export function getInlineExtensions(): Subscribable<ExecutableExtension[]> {
+    const extensionName = 'template'
+    const { manifestURL, scriptURL } = getURLsForInlineExtension('template')
+    // return fromFetch(manifestURL, undefined, response => checkOk(response).json() as Promise<ExtensionManifest>).pipe(
```
Comment (lguychard) [inline browser/src/shared/platform/inlineExtensionsService.ts:30]:
```suggestion
```

File: browser/src/shared/platform/inlineExtensionsService.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+                [
+                    {
+                        id: `sourcegraph/${extensionName}`,
+                        manifest,
+                        scriptURL,
+                    },
+                ] as ExecutableExtension[]
+        )
+
+    return from(requestPromise).pipe(shareReplay(1))
```
Comment (lguychard) [inline browser/src/shared/platform/inlineExtensionsService.ts:44]:
```suggestion
    return from(requestPromise)
```

`shareReplay()` is not necessary here, this should always replay the value of the resolved promise upon subscription.

File: browser/src/shared/platform/inlineExtensionsService.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * Determine if inline extensions should be loaded.
+ *
+ * This requires the browser extension to be built with inline extensions enabled.
+ * At build time this is determined by `shouldBuildWithInlineExtensions`.
+ */
+export const shouldUseInlineExtensions = (): boolean => isExtension && isFirefox()
+
+/**
+ * Get the manifest URL and script URL for a Sourcegraph extension which is inline (bundled with the browser addon).
```
Comment (lguychard) [inline browser/src/shared/platform/inlineExtensionsService.ts:18]:
```suggestion
 * Get the manifest URL and script URL for a Sourcegraph extension which is inline (bundled with the browser add-on).
```

File: shared/src/api/extension/worker.ts (hunk 1 above)
Comment (marekweb) [reply shared/src/api/extension/worker.ts]: Looks like a good solution, thanks.

File: shared/src/api/client/api/extensions.ts (hunk 2 above)
Comment (lguychard) [reply shared/src/api/client/api/extensions.ts:26]: I now [understand why this is happening](https://sourcegraph.slack.com/archives/CMT39K56Z/p1593174815111900). There are a few options here:
1. Modifying `getInlineExtensions()` to return an Observable that never completes: `merge(requestPromise, NEVER)`. This is not ideal, as it doesn't prevent us from running into this problem in the future.
2. Adding a comment explaining why this block is necessary
3. Changing `from(extensionRegistry.activeExtensions)` to `merge(extensionRegistry.activeExtensions, NEVER)` to ensure that, whatever was passed in, the source never completes, and explaining why we do this in a comment
4. Filtering emissions containing falsy values in the `pipe()` above, and explaining why we do this in a comment.

My preference would go to 3. or 4.

File: shared/src/api/extension/worker.ts [hunk 8]
```diff
@@ -1,14 +1,19 @@
-import ExtensionHostWorker from 'worker-loader?inline&name=extensionHostWorker.bundle.js!./main.worker.ts'
+import ExtensionHostWorker from 'worker-loader?name=extensionHostWorker.bundle.js!./main.worker.ts'
```
Comment (felixfbecker) [inline shared/src/api/extension/worker.ts:1]: We may want to avoid this import for Firefox

Comment (lguychard) [reply shared/src/api/extension/worker.ts:1]: I don't think that's necessary, what makes you think so?

Comment (lguychard) [reply shared/src/api/extension/worker.ts:1]: Elaborating: I don't think that's necessary given that this is essentially a build-time directive disguised as an import statement. It merely tells `worker-loader` to take the entrypoint `main.worker.ts`, and bundle it to a file named `extensionHostWorker.bundle.js`. Then, `new ExtensionHostWorker()` will simply be transpiled to `new Worker('extensionHostWorker.bundle.js')` (with some indirection), a statement that we'll never evaluate on Firefox, and the presence of which in the transpiled output causes no particular issue.

Comment (felixfbecker) [reply shared/src/api/extension/worker.ts:1]: Yeah you're right

File: shared/src/api/client/api/extensions.ts (hunk 2 above)
Comment (marekweb) [reply shared/src/api/client/api/extensions.ts:26]: I implemented option 3
